### DISCUSSION TITLE
v2: How do we expose functions on the window object?

### DISCUSSION THREAD

**Discussion by LukeTOBrien** [parcel-bundler/parcel#5425] [Q&A] [answered]
Hello,

I have a project that exposes functions on the `window` object that are then used by another web app.  
My functions are written as TypeScript module, and this is how I would like to expose my functions:
```
window.namespace.utils.messageBox
window.namespace.main.initMain
```
Here is my code:
index.ts
```typescript
import { Functions } from './src/Functions';
window['namespace'] = Functions;
```
Functions.ts
```typescript
import { UtilFunctions } from './Utils/UtilFunctions';
import { MainFunctions } from './Main/InitMain/MainFunctions';

export module Functions {
    export var main = MainFunctions;
    export var utils = UtilFunctions;
}
```
UtilFunctions.ts
```typescript
import { MessageBox } from './MessageBox/MessageBox';

export module UtilFunctions {
    export var messageBox = MessageBox
};
```
MainFunctions.ts
```typescript
import { InitMain } from "./InitMain";

export module MainFunctions {
    export var initMain = InitMain;
}
```

It was working perfectly with Parcel 1, all I required was a `.terserrc`:
```
.terserrc:
{
  "mangle" false
}
```

With Parcel 2, I can open up the built bundle and see that `window['namespace'] = Functions` has been transformed to be `window.namespace = ...`.  
But when I use the F12 Dev Tools, `window.namespace` is undefined, so if seems to me that in the bundle perhap a variable called `window` is at function scope and it is not the global window object.

Do you see what I want to achieve here?  
Does anyone know how to do this in Parcel 2?

Thanks.

**Comment by mischnic** [comment]
Please provide a full example, this is working for me:

```js
// index.ts
import { Foo } from "./other";
window["namespace"] = Foo;

// other.ts
export const Foo = {
	a: 1,
	b: 2,
};
```
```html
<!-- index.html -->
<script src="index.ts"></script>
```

And printing `window.namespace` in the devtools works as expected

**Reply by LukeTOBrien** [reply]
I have created a [lil repo here](https://github.com/LukeTOBrien/ParcelNamespace).  
I am using `modules` which cannot be used with Babel (get an experiimental error, I am using `@parcel/transformer-typescript-tsc@beta-1`

```ts
export function Hello() {
    console.log("Hello");
}

export function Bye() {
    console.log("Bye");
}
```
...
```ts
import { Bye, Hello } from "./MoreFunctions";

export module Functions {
    export var hello = Hello;
    export var bye = Bye;
}
```
...
```ts
import { Functions } from "./Functions";

window['namespace'] = Functions;
```

In console:
```
Uncaught TypeError: window.namespace is undefined
```

Works fine in Parcel 1, could it be `transformer-typescript-tsc`?

**Reply by mischnic** [reply]
Both when serving the `out` folder in your repo and after running `yarn build`, this happens for me in the browser: "Hello" is printed and an alert with "undefined" opens.

**Comment by LukeTOBrien** [comment, marked as answer]
Sorry my bad... The script was not loading due to a `/` at the begining.  
I am on Windows, perhaps the file system loads it differently?

```html
<script src="src/script.ts"></script>
```
->
```html
<script src="/script.39e78319.js"></script>
```

I removed the `/` and it loaded fine.... Thanks you can close this issue.
